crates/wit-bindgen-rust/src/interface.rs

```diff
@@ -316,42 +305,148 @@ fn {method_name}(
                 self.print_ty(ty, TypeMode::owned());
                 self.push_str(",");
             }
-            uwrite!(
+
+            uwriteln!(
                 self.src,
                 r#"),
         Tx>) {{
             use {tracing}::Instrument as _;
             let span = {tracing}::trace_span!("{method_name}");
             let _enter = span.enter();
             let handler = self.clone();
-            {tokio}::spawn(async move {{
-                match handler.{name}(context, "#,
-                name = to_rust_ident(name),
+                {tokio}::spawn(async move {{"#,
                 tokio = self.gen.tokio_path(),
-                tracing = self.gen.tracing_path(),
+                tracing = self.gen.tracing_path()
             );
+
+            // print function call
+            match kind {
+                FunctionKind::Freestanding => {
+                    uwrite!(
+                        self.src,
+                        "match handler.{name}(context, ",
+                        name = to_rust_ident(name)
+                    );
+                }
+                FunctionKind::Constructor(id) => {
+                    let resource_name = self.resolve.types[*id]
+                        .name
+                        .as_ref()
+                        .unwrap()
+                        .to_upper_camel_case();
+
+                    uwrite!(
+                        self.src,
+                        "match <T as Handler<Ctx>>::{resource_name}::new(context, ",
+                    )
+                }
+                FunctionKind::Method(id) => {
```

Comment:
We should probably reuse https://github.com/wrpc/wrpc/blob/a75b04391524217058fb4c5284aee224ee8e4464/crates/wit-bindgen-go/src/interface.rs#L28-L41 here, I'll work on setting up a `wit-bindgen-wrpc-core` crate, which would contain such kind of reusable utilities

Reply:
Awesome, that would be very nice!

Reply:
There's now `wrpc-introspect` crate available, which exports `rpc_func_name` https://wrpc.github.io/wrpc/wrpc_introspect/fn.rpc_func_name.html